add onnx graph-optimization library

anderspapitto opened the pull request.

bddppq commented on onnx_caffe2/bin/conversion.py:
```diff
@@ -79,7 +79,8 @@ def caffe2_to_onnx(caffe2_net,
 def onnx_to_caffe2(onnx_model, output, init_net_output):
     onnx_model_proto = ModelProto()
-    onnx_model_proto.ParseFromString(onnx_model.read())
+    onnx_model_proto.ParseFromString(c2.optimize_onnx(onnx_model.read()))
```
There are two ways of using onnx-caffe2, one is do file to file conversion (here), another is use the python interface to run an onnx model via caffe2 on the fly (which is in backend.py). Both will go through the same translation path `onnx_graph_to_caffe2_net`, so this `optimize_onnx` should be injected into there.

anderspapitto replied:
that can probably be done, but it requires a more extensive refactoring because

- onnx_graph_to_caffe2_net operates on Graphs, and optimize-onnx needs the whole Model
- process() does destructive things before calling onnx_graph_to_caffe2_net, like stripping out the initializers

dzhulgakov commented on manipulation/src/utils/auto_gpu.h:
```diff
@@ -0,0 +1,69 @@
+#include <cuda_runtime.h>
+#endif
+
+struct AutoGPU {
```
we probably don't need gpu access or anything related to it

anderspapitto replied:
sure, I'll go through and clear more of that stuff out

bddppq commented:
Please create a "third_party" directory and put all submodules into it.

dzhulgakov commented on manipulation/src/onnx.pb.cc:
```diff
@@ -0,0 +1,8946 @@
+// Generated by the protocol buffer compiler.  DO NOT EDIT!
```
Having yet another checked-in copy of protobuf is messy. Can we depend on generated protoc files from onnx binary installation or just invoke protoc as a part of cmake script directly?

anderspapitto replied:
probably - this was the quickest way to get things working, but I'll try one of those approaches

anderspapitto commented:
++ for third_party

anderspapitto commented:
addressed @jamesr66a comments from the PR in the other repo. DIdn't address the other comments yet.

anderspapitto commented:
addressed all comments except

- not adding a copy of generated protobuf files
- moving entry point into onnx_graph_to_caffe2

anderspapitto commented:
protoc now runs during build, and onnx/onnx is submoduled to gain access to the .proto file

anderspapitto commented:
moved optimize_onnx() into onnx_graph_to_caffe2(), so all comments are now addressed

@bddppq do you actually think this is preferable? I had to substantially shuffle around the innards of backend.py

bddppq commented on onnx_caffe2/backend.py:
```diff
@@ -648,31 +659,33 @@ def onnx_graph_to_caffe2_net(cls, graph_def):
 
-        predict_net.external_input.extend(
-            value_info.name for value_info in graph_def.input)
+        if include_inputs_and_initializers:
```
This part should not be skipped, otherwise the predict_net would be invalid (inner nodes refer to nonexistent input blobs).

bddppq commented on onnx_caffe2/backend.py:
```diff
@@ -631,9 +639,12 @@ def _all_names_in_graph(graph):
 
     @classmethod
-    def onnx_graph_to_caffe2_net(cls, graph_def):
+    def onnx_graph_to_caffe2_net(cls, model, include_inputs_and_initializers):
```
rename this, it's not taking graph as input anymore

dzhulgakov commented on .travis/build.sh:
```diff
@@ -3,10 +3,12 @@
 
 onnx_c2_dir="$PWD"
-pip install -e $onnx_c2_dir
+CC=gcc-5 CXX=g++-5 pip install -e $onnx_c2_dir
```
why gcc5?

bddppq replied:
@dzhulgakov pytorch gcc4.8 build is currently broken.
@anderspapitto you probably also need to configure travis to install gcc5

anderspapitto replied:
yeah, I have modified the travis setup in this diff.

(specifically, it's aten which doesn't build with gcc4.8

dzhulgakov commented on manipulation/src/export.cpp:
```diff
@@ -0,0 +1,233 @@
+  TensorType* node_type = n->type().get();
+  const std::vector<Dimension>& dims = node_type->sizes();
+  for (const Dimension& d : dims) {
+    dd = &d;
```
not sure where it came from, but it's weird

anderspapitto replied:
++

dzhulgakov commented on manipulation/src/export.cpp:
```diff
@@ -0,0 +1,233 @@
+
+void encodeGraph(onnx::GraphProto * p_g, const std::shared_ptr<Graph> & g) {
+  JIT_ASSERT(p_g != nullptr);
+  p_g->set_name("torch-jit-export");
```
let's preserve the original graph name

dzhulgakov commented on manipulation/src/test.py:
```diff
@@ -0,0 +1,29 @@
+    input_model = onnx.load(input_proto_path)
+    output_model = onnx.load(output_proto_path)
+
+    assert onnx.helper.printable_graph(input_model.graph) \
```
it might be better to compare source protobufs (e.g. str(graph_a) == str(graph_b)) directly as printable_graph() doesn't print all fields.

bddppq replied:
I think you can directly do `graph_a == graph_b`

anderspapitto replied:
I can do this, but I'll have to add more code to preserve comments everywhere (almost every onnx struct has a comment field that I'm currently not preserving)

dzhulgakov commented on manipulation/src/test.py:
```diff
@@ -0,0 +1,29 @@
+    assert onnx.helper.printable_graph(input_model.graph) \
+        == onnx.helper.printable_graph(output_model.graph)
+
+def test_torch_model(roundtrip_binary):
```
not invoked?

dzhulgakov commented on manipulation/src/test.py:
```diff
@@ -0,0 +1,29 @@
+    torch.onnx.export(model, x, input_path)
+    roundtrip_helper(roundtrip_binary, input_path, output_path)
+
+# invoke as
```
maybe turn it to unittest and invoke some of the models from the model zoo?

anderspapitto replied:
++

dzhulgakov commented on manipulation/src/main.cpp:
```diff
@@ -0,0 +1,21 @@
+  mp.ParseFromString(buffer.str());
+  std::shared_ptr<onnx::optimization::Graph> g = onnx::optimization::ImportModel(mp);
+
+  std::string s = onnx::optimization::ExportGraph(g);
```
you probably want to preserve original metadata from ModelProto

dzhulgakov commented on onnx_caffe2/backend.py:
```diff
@@ -472,6 +473,12 @@ def _direct_initialize_inputs(cls, inputs, initialized, ws, device_option):
 
+    @staticmethod
+    def optimize_onnx(input):
+        proc = Popen(['optimize-onnx'], stdin=PIPE, stdout=PIPE, stderr=PIPE)
```
should it be a relative path?

anderspapitto replied:
I've changed setup.py to put `optimize-onnx` in the bin/ of the python install/virtualenv, so it should be fine to look for it in PATH

bddppq replied:
@anderspapitto This is not always true, you can use the virtualenv without activating the virtualenv (thus `virtualenv/bin` is not in your `$PATH`).

dzhulgakov commented on onnx_caffe2/backend.py:
```diff
@@ -629,10 +624,14 @@ def _all_names_in_graph(graph):
 
     @classmethod
-    def onnx_graph_to_caffe2_net(cls, graph_def, device="CPU"):
+    def onnx_model_to_caffe2_net(cls, model, include_inputs_and_initializers, device="CPU"):
```
make it the last argument (as some callsites might pass device already) and default to true

bddppq replied:
the function name has been changed, this will be a breaking change anyway (but to be fair this function is never meant to be a public function)

dzhulgakov replied:
it'd actually invoked (we should use '_' prefix for private stuff): https://fburl.com/8z9yn2ck and https://github.com/ezyang/onnx-pytorch/search?utf8=%E2%9C%93&q=onnx_graph_to_caffe2_net&type=

bddppq commented on setup.py:
```diff
@@ -57,15 +59,38 @@ def run(self):
+class install(setuptools.command.install.install):
+    def run(self):
+        setuptools.command.install.install.run(self)
+        source = os.path.join(os.getcwd(), 'build/manipulation/src/optimize-onnx')
```
for the purpose of cross platform, use `os.path.join` to concat each component

bddppq commented on setup.py:
```diff
@@ -57,15 +59,38 @@ def run(self):
         self.run_command('create_version')
+        self.run_command('build')
         setuptools.command.develop.develop.run(self)
+        source = os.path.join(os.getcwd(), 'build/manipulation/src/optimize-onnx')
```
Probably don't rely on `getcwd`, use `__file__`?

anderspapitto replied:
++

anderspapitto commented on tests/roundtrip_optimize_onnx.py:
```diff
@@ -0,0 +1,67 @@
+
+    # arbitrarily pick one relatively small model to sanity test with
+    def test_squeezenet(self):
+        self._roundtrip('squeezenet-ir-version-3')
```
note: this passes locally, but won't pass in CI until I'm able to upload the new ir version of this model

dzhulgakov commented on optimize-onnx/src/main.cpp:
```diff
@@ -0,0 +1,58 @@
+  std::shared_ptr<onnx::optimization::Graph> g = onnx::optimization::ImportModel(mp_in);
+
+  if (g.get() == nullptr) {
+    std::cout << buffer.str();
```
print some warning to stderr?

dzhulgakov commented on optimize-onnx/src/main.cpp:
```diff
@@ -0,0 +1,58 @@
+// doc_strings are lost, but semantics should be unchanged.
+int main(int argc, char** argv) {
+  std::stringstream buffer;
+  buffer << std::cin.rdbuf();
```
this might a problem with reading binary data: https://stackoverflow.com/questions/7587595/read-binary-data-from-stdcin or does rdbuf() solve all of that?

dzhulgakov commented on onnx_caffe2/backend.py:
```diff
@@ -627,10 +631,23 @@ def _all_names_in_graph(graph):
+    def _onnx_model_to_caffe2_net(cls, model, device, include_inputs_and_initializers):
         device_option = get_device_option(Device(device))
+
+        # before_opt = model.SerializeToString()
```
remove?

anderspapitto replied:
++

bddppq commented:
We haven't done any binary distribution for onnx packages so far. With the addition of c++ component, bdist would definitely be more difficult, let's just use source distribution for now.

ezyang commented:
We have conda-forge packaging for onnx-caffe2 now (https://github.com/conda-forge/onnx-caffe2-feedstock) so binary distribution should be easy.

anderspapitto commented on optimize-onnx/src/import.cpp:
```diff
@@ -0,0 +1,252 @@
+    inputs.reserve(np.input_size());
+    for (int i = 0; i < np.input_size(); i++) {
+      inputs.push_back(np.input(i));
+      inputs_by_node[n] = inputs;
```
++

anderspapitto commented:
Yes, I need to rebase and finish fixing one (known) bug (I always rely on raw_data, rather than float_data/int_data/string_data, etc).

As far as I know it should not be a big problem for binary distributions, unless they have difficulty with native executables. I can test that it's compatible with whatever solutions we currently have (e.g. the conda package).

anderspapitto commented:
rebased, added support for all the various <datatype>_data fields in TensorProto, and along the way largely removed the ATen dependency

dzhulgakov commented on onnx_caffe2/backend.py:
```diff
@@ -501,6 +504,20 @@ def _direct_initialize_inputs(cls, inputs, initialized, ws, device_option):
 
+    @staticmethod
+    def optimize_onnx(input):
+        # this is where the binary is located when onnx-caffe2 has been installed
```
you might need to modify it for internal BUCK-based build. Probably, try to import optimizer as a python module if it's installed and otherwise - fallback on the binary.

anderspapitto replied:
ok, I'll be aware of that for when it comes up

anderspapitto commented:
now that it's approved, I've rebased and squashed most of the commits down. Assuming tests still pass someone with write access can merge it.
